allennlp/models/coreference_resolution/coref.py

```diff
@@ -0,0 +1,244 @@
+        combined_idx = use_identity * distances + (1 + (-1 * use_identity)) * logspace_idx
+        return combined_idx.clamp(0, 9)
+
+    def forward(self,  # type: ignore
+                text: Dict[str, torch.LongTensor],
+                span_starts: torch.IntTensor,
+                span_ends: torch.IntTensor,
+                span_labels: torch.IntTensor,
+                metadata: List[Dict[str, Any]] = None) -> Dict[str, torch.Tensor]:
+        # pylint: disable=arguments-differ
+        # Basic setup.
+        text_emb = self._lexical_dropout(self._text_field_embedder(text)) # [b, text_len, emb]
+        is_cuda = text_emb.is_cuda
+        batch_size = span_labels.size(0)
+        num_spans = span_labels.size(1)
+        text_len = text_emb.size(1)
+        text_mask = util.get_text_field_mask(text).float() # [b, text_len]
+        span_mask = (span_starts >= 0).float() # [b, num_spans, 1]
+        span_starts = F.relu(span_starts.float()).long() # [b, num_spans, 1]
+        span_ends = F.relu(span_ends.float()).long() # [b, num_spans, 1]
+
+        # Compute span representations.
+        contextualized_emb = self._context_layer(text_emb, text_mask) # [b, text_len, emb]
+        start_emb = util.batched_index_select(contextualized_emb,
+                                              span_starts.squeeze(-1)) # [b, num_spans, emb]
+        end_emb = util.batched_index_select(contextualized_emb,
+                                            span_ends.squeeze(-1)) # [b, num_spans, emb]
+        span_size = span_ends - span_starts # [b, num_spans, 1] (really span size minus one)
+        span_size_emb = self._span_width_embedding(span_size.squeeze(-1)) # [b, num_spans, emb]
+        head_scores = self._head_scorer(contextualized_emb) # [b, text_len, 1]
+        head_offsets = util.get_indices(self._max_span_width, is_cuda).view(1, 1, -1) # [1, 1, max_span_width]
+        head_mask = (head_offsets <= span_size).float() # [b, num_spans, max_span_width]
+        raw_head_idx = span_ends - head_offsets # [b, num_spans, max_span_width]
+        head_mask = head_mask * (raw_head_idx >= 0).float() # [b, num_spans, max_span_width]
+        head_idx = F.relu(raw_head_idx.float()).long() # [b, num_spans, max_span_width]
+        flat_head_idx = util.flatten_batched_indices(head_idx,
+                                                     text_len) # [b * num_spans * max_span_width]
+        span_text_emb = util.batched_index_select(text_emb,
+                                                  head_idx,
+                                                  flat_head_idx) # [b, num_spans, max_span_width, emb]
+        span_head_scores = util.batched_index_select(head_scores,
+                                                     head_idx,
+                                                     flat_head_idx).squeeze(-1) # [b, num_spans, max_span_width]
+        span_head_scores += head_mask.float().log() # [b, num_spans, max_span_width]
```

Comment:
IMO, this is a lot cleaner and faster than what's in masked_softmax. Masking in logspace only requires a single elementwise multiply. In practice there are no numerical stability issues. I was actually going to add a pull request that would fix this in bidaf...

Reply:
Would it make sense to just fix our `masked_softmax`? I never thought of doing it like this, but it seems reasonable as long as it doesn't give NaNs (I'm still a little surprised it works, because you're taking `log(0)`...).

The benefit of using the function is that it improves readability of the code - you just have one line with a very transparent function call, instead of four lines of less-transparent logic.

Reply:
Not sure an extra abstraction is necessary here. Is `util.mask_softmax(scores, mask)` really more readable than `F.softmax(scores + mask.log())`? There are also a lot of cases where you'd want to reuse the logged masks or the masked scores, and chances are you'll end up with inconsistent usage of this API for efficiency reasons.